crates/core/src/network/io/reactor.rs

```diff
@@ -0,0 +1,465 @@
+                (State::Active, Decision::Drain(reason)) => {
+                    // send drain signal, and switch
+                    self.switch_to_draining(reason, &peer_router);
+                }
+                (State::Active, Decision::NotifyPeerShutdown) => {
+                    conn_tracker.notify_peer_shutdown(self.connection.peer());
+                    self.switch_to_draining(DrainReason::ConnectionDrain, &peer_router);
+                }
+                (_, Decision::DrainEgress) => {
+                    self.tx.take();
+                    self.state = State::WaitForEgress;
+                }
+                (State::Draining { .. }, Decision::NotifyPeerShutdown) => {
+                    conn_tracker.notify_peer_shutdown(self.connection.peer());
+                }
+                (State::Draining { .. }, Decision::Drain(reason)) => {
+                    // send drain signal, and switch
+                    self.send_drain_signal(reason);
+                }
+                (_, Decision::Drop) => break,
+                (State::WaitForEgress, Decision::NotifyPeerShutdown) => unreachable!(),
+                (State::WaitForEgress, Decision::Drain(_)) => unreachable!(),
+            }
+        }
+
+        // we also try to deregister here because we don't always transition to "Draining" before
+        // drppping
+        peer_router.deregister(&self.connection);
+        conn_tracker.connection_dropped(&self.connection);
+        Ok(())
+    }
+
+    async fn handle_message(
+        &mut self,
+        msg: Option<Result<Message, ProtocolError>>,
+        router: &impl Handler,
+    ) -> Decision {
+        let msg = match msg {
+            Some(Ok(msg)) => msg,
+            Some(Err(status)) => {
+                // This indicates an unrecoverable error in the underlying stream.
+                // It's very likely that we can't send or receive anything.
+                // We'll choose to drop the queued messages on the floor.
+                info!("Error received: {status}, terminating stream");
+                return Decision::Drop;
+            }
+            None if self.state.is_active() => {
+                return Decision::Drain(DrainReason::ConnectionDrain);
+            }
+            None => {
+                // Peer terminated the connection stream has terminated cleanly.
+                // no more messages will arrive on this stream.
+                return Decision::DrainEgress;
+            }
+        };
+
+        let processing_started = Instant::now();
+        // body are not allowed to be empty.
+        let Some(body) = msg.body else {
+            return Decision::Drain(DrainReason::CodecError(
+                "Body is missing on message".to_owned(),
+            ));
+        };
+
+        //  header is required on all messages
+        let Some(header) = msg.header else {
+            warn!("Peer sent a message without header");
+            return Decision::Drop;
+        };
```

Comment:
Out of curiosity, why are we taking different decisions here? Don't both cases signify some kind of protocol violation?

Reply:
My intuition here is that codec errors could happen due to version incompatibility, so we might still be able to communicate the error back and drain the connection gracefully. The header is an invariant that we won't remove.